### PR TITLE
Mss 952/apple update subscription lambda

### PR DESCRIPTION
**Pull request by NathanielBennett**
Implements storage of apple subscriptions: Needs https://github.com/guardian/mobile-purchases/pull/74 to be merged first ( this is where the dynamo cfn changes come from )

### CODE REVIEW

File: typescript/src/updatesubs/apple.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        await parseAndStoreSubscriptionUpdate(record, makeDynamoPromise)
+    })
+
+    return Promise.all(emptyPromises)
+        .then( value => {
+            console.log(`Processed ${event.Records.length} subscriptions`)
+            "OL"
+        })
+        .catch(error => {
+            console.error("Error processing subsctption update: ", error)
```
Comment (davidfurey) [inline typescript/src/updatesubs/apple.ts]: Typo "subsctption"

File: typescript/src/updatesubs/apple.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+export async function handler(event: SQSEvent) {
+    const emptyPromises = event.Records.map( async (record) => {
+        await parseAndStoreSubscriptionUpdate(record, makeDynamoPromise)
+    })
+
+    return Promise.all(emptyPromises)
+        .then( value => {
+            console.log(`Processed ${event.Records.length} subscriptions`)
+            "OL"
```
Comment (davidfurey) [inline typescript/src/updatesubs/apple.ts]: What does "OL" mean?

File: typescript/src/updatesubs/google.ts [hunk 3]
```diff
@@ -22,7 +26,23 @@ interface GoogleResponseBody {
 
 const restClient = new restm.RestClient('guardian-mobile-purchases');
 
-export function getGoogleSubResponse(record: SQSRecord): Promise<SubscriptionUpdate> {
+function parseAutoRenewing(autoRenewing: string | undefined) : boolean | undefined {
+
+        if(autoRenewing) {
+            try {
+                return JSON.parse(autoRenewing)
```
Comment (davidfurey) [inline typescript/src/updatesubs/google.ts]: Seems like overkill to use JSON.parse to check if a string is true or false

File: typescript/src/updatesubs/updatesub.ts [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
     if (cancellationTime) {
         return new Date(Number.parseInt(cancellationTime)).toISOString()
     } else {
         return ""
     }
 }
 
-function makeTimeToLive(date: Date) {
+export function makeTimeToLive(date: Date) {
     return Math.ceil((date.getTime() / 1000) + 7 * ONE_YEAR_IN_SECONDS)
```
Comment (davidfurey) [inline typescript/src/updatesubs/updatesub.ts]: Maybe not one for this PR, but I don't think we should be keeping this data for 7 years

File: typescript/src/models/subscription.ts [hunk 5]
```diff
@@ -19,23 +20,49 @@ export class Subscription {
...
-    googlePayload?: any;
-
     @attribute()
     ttl?: number;
 
-    constructor(subscriptionId: string, startTimeStamp?: string, endTimeStamp?: string, cancellationTimeStamp?: string, autoRenewing?: boolean, googlePayload?: any,  ttl?: number) {
+    constructor(subscriptionId: string, startTimeStamp?: string, endTimeStamp?: string, cancellationTimeStamp?: string, autoRenewing?: boolean, ttl?: number) {
         this.subscriptionId = subscriptionId;
-        this.startTimeStamp = startTimeStamp;
+        this.startTimeStamp = startTimeStamp
```
[the remark this answers is not in the document]
Comment (NathanielBennett) [reply typescript/src/models/subscription.ts]: Good point. Not yet. Although ts semis are a bit like scala ones, right?

File: typescript/src/updatesubs/google.ts (hunk 3 above)
Comment (NathanielBennett) [reply typescript/src/updatesubs/google.ts]: Actually it doesn't really matter because It's not actually used. Binning it.

File: typescript/src/updatesubs/updatesub.ts (hunk 4 above)
Comment (NathanielBennett) [reply typescript/src/updatesubs/updatesub.ts]: OK, that was the original requirement. See @alexduf doxs.

File: cloudformation.yaml [hunk 6]
```diff
@@ -668,6 +668,35 @@ Resources:
...
+      FunctionName: !Sub ${App}-apple-update-subscriptions-${Stage}
+      Code:
+        S3Bucket: !Ref DeployBucket
+        S3Key: !Sub ${Stack}/${Stage}/${App}-apple-update-subscriptions/apple-update-subscriptions.zip
+      Environment:
+        Variables:
+          App: !Sub ${App}
+          Stack: !Sub ${Stack}
+          Stage: !Sub ${Stage}
+      Description: Consomes subscription data updates from app store from sqs and stores them in dynamo
```
Comment (alexduf) [inline cloudformation.yaml]:
```suggestion
      Description: Consumes subscription data updates from app store from sqs and stores them in dynamo
```

File: typescript/src/updatesubs/apple.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    receipt: string,
+    transactionId: string,
+    startDate: string,
+    cancellationDate: string,
+    endDate: string,
+    autoRenewing: boolean
+    appleBody: string
+}
+
+export function makeDynamoPromise(record: SQSRecord): Promise<AppleSubscription>  {
```
Comment (alexduf) [inline typescript/src/updatesubs/apple.ts]: I think this method it misnamed, as currently it converts  a model to another model.

Now, I think we need to either verify the cryptographic signature that the receipt at the pubsub stage, or have an API call to Apple at this point to check the receipt is valid and up to date.

That might be a good case for a separate PR?

File: typescript/src/updatesubs/apple.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        makeCancellationTime(sub.cancellationDate),
+        sub.autoRenewing,
+        sub.appleBody,
+        makeTimeToLive(new Date(Date.now()))
+    ))
+}
+
+export async function handler(event: SQSEvent) {
+    const emptyPromises = event.Records.map( async (record) => {
+        await parseAndStoreSubscriptionUpdate(record, makeDynamoPromise)
```
Comment (alexduf) [inline typescript/src/updatesubs/apple.ts]: here I think you need a return statement because of the curly braces

File: typescript/src/updatesubs/apple.ts (hunk 7 above)
Comment (NathanielBennett) [reply typescript/src/updatesubs/apple.ts]: Agreed. Do we have details of he apple API call to make. Lets do this as a separate PR at the pubsub test

**Comment by NathanielBennett** [comment]
I've created a story to validate the reciept: https://theguardian.atlassian.net/browse/MSS-1054
